externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLLineageParseHelper.scala

```diff
@@ -0,0 +1,366 @@
+
+      case p: HiveTableRelation =>
+        val tableName = p.tableMeta.qualifiedName
+        val dataCols = p.dataCols.map(_.exprId.id.toString).toSet
+        val dataColNames = p.output.map(_.name)
+        if (parentColumnsLineage.nonEmpty) {
+          parentColumnsLineage.map { case (k, columns) =>
+            k -> columns.collect {
+              case col if dataCols.contains(col.split('#').last) => s"$tableName.$col"
+            }
+          }
+        } else {
+          val colPair = dataColNames.map { col =>
+            col -> List(s"$tableName.$col")
+          }
+          ListMap(colPair: _*)
+        }
+
+      case p: DataSourceV2Relation =>
+        val tableName = p.name
+        val dataCols = p.output.map(_.exprId.id.toString).toSet
+        val dataColNames = p.output.map(_.name)
+        if (parentColumnsLineage.nonEmpty) {
+          parentColumnsLineage.map { case (k, columns) =>
+            k -> columns.collect {
+              case col if dataCols.contains(col.split('#').last) => s"$tableName.$col"
+            }
+          }
+        } else {
+          val colPair = dataColNames.map { col =>
+            col -> List(s"$tableName.$col")
+          }
+          ListMap(colPair: _*)
+        }
+
+      case p: DataSourceV2ScanRelation =>
+        val tableName = p.name
+        val dataCols = p.output.map(_.exprId.id.toString).toSet
+        val dataColNames = p.output.map(_.name)
+        if (parentColumnsLineage.nonEmpty) {
+          parentColumnsLineage.map { case (k, columns) =>
+            k -> columns.collect {
+              case col if dataCols.contains(col.split('#').last) => s"$tableName.$col"
+            }
+          }
+        } else {
+          val colPair = dataColNames.map { col =>
+            col -> List(s"$tableName.$col")
+          }
+          ListMap(colPair: _*)
+        }
+
+      case p: LocalRelation =>
+        val dataCols = p.output.map(_.exprId.id.toString).toSet
+        val dataColNames = p.output.map(_.name)
+        if (parentColumnsLineage.nonEmpty) {
+          parentColumnsLineage.map { case (k, columns) =>
+            k -> columns.collect {
+              case col if dataCols.contains(col.split('#').last) => col
+            }
+          }
+        } else {
+          val colPair = dataColNames.map { col =>
+            col -> List(col)
+          }
+          ListMap(colPair: _*)
+        }
+
+      case p if p.children.isEmpty => ListMap.empty
+      case p =>
+        p.children.map(extractColumnsLineage(_, parentColumnsLineage)).reduce(mergeColumnsLineage)
+    }
+  }
+
+  private def getSelectColumnLineage(exps: Seq[NamedExpression]): ListMap[String, List[String]] = {
+    val result = exps.map { exp =>
+      val derivedColumns = exp.references.map(e => s"${e.name}#${e.exprId.id}").toList
+      exp match {
+        case e: Alias =>
+          (s"${e.name}#${e.exprId.id}" -> derivedColumns)
+        case _ =>
+          (exp.toString -> derivedColumns)
+      }
```

Comment:
can we use `AttributeMap` instead of `ListMap[String, List[String]]` ?  for example:

```scala
  def mergeReference(
      left: AttributeMap[AttributeSet],
      right: AttributeMap[AttributeSet]): AttributeMap[AttributeSet] = {
    AttributeMap(
      left ++ right.map { case (k, attrs) =>
        k -> (attrs ++ right.getOrElse(k, AttributeSet.empty))
      }
    )
  }

  def replaceParentReferenceByChildAlias(
      parent: AttributeMap[AttributeSet],
      child: AttributeMap[AttributeSet]): AttributeMap[AttributeSet] = {
    AttributeMap(
      parent.map { case (k, attrs) =>
        k -> AttributeSet(attrs.flatMap(attr => child.get(attr).getOrElse(AttributeSet(attr))))
      }
    )
  }

  private def selectAttributeReferenceMap(
      named: Seq[NamedExpression]): AttributeMap[AttributeSet] = {
    AttributeMap.apply(
      named.map {
        case a: Alias => a.toAttribute -> a.references
        case a: Attribute => a -> a.references
      }
    )
  }

```